File: rust/src/domain/spectrogram.rs
#![allow(dead_code)]

use crate::domain::fft_processor::{FftProcessor, WindowType};
use crate::domain::Waveform;

#[derive(Debug, Clone)]
pub struct SpectrogramConfig {
    pub window_size: usize,
    pub overlap: usize,
    pub min_freq: f32,
    pub max_freq: f32,
}

impl Default for SpectrogramConfig {
    fn default() -> Self {
        Self {
            window_size: 1024,
            overlap: 512,
            min_freq: 0.0,
            max_freq: 22050.0,
        }
    }
}

#[derive(Debug, Clone)]
pub struct SpectrogramData {
    pub frequencies: Vec<f32>,
    pub time_bins: Vec<i64>,
    pub magnitudes: Vec<Vec<f32>>,
    pub sample_rate: u32,
    pub window_size: usize,
    pub overlap: usize,
}

impl SpectrogramData {
    pub fn new(
        frequencies: Vec<f32>,
        time_bins: Vec<i64>,
        magnitudes: Vec<Vec<f32>>,
        sample_rate: u32,
        window_size: usize,
        overlap: usize,
    ) -> Self {
        Self {
            frequencies,
            time_bins,
            magnitudes,
            sample_rate,
            window_size,
            overlap,
        }
    }

    pub fn frequency_bin(&self, freq: f32) -> Option<usize> {
        self.frequencies
            .iter()
            .position(|&f| (f - freq).abs() < 1.0)
    }

    pub fn time_bin(&self, timestamp_ms: i64) -> Option<usize> {
        self.time_bins
            .iter()
            .position(|&t| (t - timestamp_ms).abs() < 1)
    }

    pub fn magnitude_at(&self, freq: f32, timestamp_ms: i64) -> Option<f32> {
        let f_idx = self.frequency_bin(freq)?;
        let t_idx = self.time_bin(timestamp_ms)?;
        Some(self.magnitudes[t_idx][f_idx])
    }

    pub fn max_magnitude(&self) -> f32 {
        self.magnitudes
            .iter()
            .flat_map(|row| row.iter())
            .cloned()
            .fold(f32::NEG_INFINITY, f32::max)
    }

    pub fn min_magnitude(&self) -> f32 {
        self.magnitudes
            .iter()
            .flat_map(|row| row.iter())
            .cloned()
            .fold(f32::INFINITY, f32::min)
    }

    pub fn to_image_data(&self, width: usize, height: usize) -> Vec<u8> {
        if self.magnitudes.is_empty() || self.frequencies.is_empty() {
            return vec![0u8; width * height * 3];
        }

        let max_mag = self.max_magnitude();
        let min_mag = self.min_magnitude();
        let range = (max_mag - min_mag).max(0.001);

        let mut pixels = Vec::with_capacity(width * height * 3);
        let time_bins_count = self.time_bins.len();
        let freq_bins_count = self.frequencies.len();

        for y in 0..height {
            let freq_idx = (y as f32 / height as f32 * freq_bins_count as f32) as usize;
            let freq_idx = freq_idx.min(freq_bins_count - 1);

            for x in 0..width {
                let time_idx = (x as f32 / width as f32 * time_bins_count as f32) as usize;
                let time_idx = time_idx.min(time_bins_count - 1);

                let magnitude = self.magnitudes[time_idx][freq_idx];
                let normalized = ((magnitude - min_mag) / range * 255.0) as u8;
                
                let r = normalized;
                let g = normalized.saturating_sub(50);
                let b = normalized.saturating_sub(100);

                pixels.push(r);
                pixels.push(g);
                pixels.push(b);
            }
        }

        pixels
    }
}

pub struct SpectrogramProcessor {
    fft_processor: FftProcessor,
}

impl SpectrogramProcessor {
    pub fn new() -> Self {
        Self {
            fft_processor: FftProcessor::new(),
        }
    }

    pub fn compute(&mut self, waveform: &Waveform, sample_rate: u32, config: SpectrogramConfig) -> SpectrogramData {
        let samples = &waveform.samples;
        let window_size = config.window_size;
        let overlap = config.overlap;
        let hop_size = window_size.saturating_sub(overlap);

        if samples.len() < window_size {
            return SpectrogramData::new(
                vec![],
                vec![],
                vec![],
                sample_rate,
                window_size,
                overlap,
            );
        }

        let start_time_ns = waveform.timestamp.timestamp_nanos_opt().unwrap_or(0);
        let mut time_bins = Vec::new();
        let mut magnitudes = Vec::new();
        let mut frequencies: Vec<f32> = Vec::new();
        let mut position = 0isize;

        while position + window_size as isize <= samples.len() as isize {
            let start = position as usize;
            let end = start + window_size;
            let window = &samples[start..end];

            let spectrum = self.fft_processor.compute_spectrum(
                window,
                sample_rate as f32,
                WindowType::Hann,
            );

            let freq_range_start = spectrum.frequencies
                .iter()
                .position(|&f| f >= config.min_freq)
                .unwrap_or(0);
            let freq_range_end = spectrum.frequencies
                .iter()
                .rposition(|&f| f <= config.max_freq)
                .map(|p| p + 1)
                .unwrap_or(spectrum.frequencies.len());

            let filtered_magnitudes: Vec<f32> = spectrum.magnitudes_db
                [freq_range_start..freq_range_end]
                .to_vec();

            if frequencies.is_empty() {
                frequencies = spectrum.frequencies
                    [freq_range_start..freq_range_end]
                    .to_vec();
            }

            let timestamp_ms = (start_time_ns / 1_000_000) + (position * 1000 / sample_rate as isize) as i64;
            time_bins.push(timestamp_ms);
            magnitudes.push(filtered_magnitudes);

            position += hop_size as isize;
        }

        SpectrogramData::new(
            frequencies,
            time_bins,
            magnitudes,
            sample_rate,
            window_size,
            overlap,
        )
    }

    pub fn compute_default(&mut self, waveform: &Waveform, sample_rate: u32) -> SpectrogramData {
        self.compute(waveform, sample_rate, SpectrogramConfig::default())
    }
}

impl Default for SpectrogramProcessor {
    fn default() -> Self {
        Self::new()
    }
}

impl SpectrogramConfig {
    pub fn with_window_size(mut self, size: usize) -> Self {
        self.window_size = size;
        self
    }

    pub fn with_overlap(mut self, overlap: usize) -> Self {
        self.overlap = overlap;
        self
    }

    pub fn with_frequency_range(mut self, min: f32, max: f32) -> Self {
        self.min_freq = min;
        self.max_freq = max;
        self
    }
}

File: sdk/dsp/src/spectrogram.cpp
// Spectrogram — ported from rust/src/domain/spectrogram.rs.
// Adapted to take raw float samples + a start timestamp (the Rust version took
// a `Waveform` entity; the C++ core is agnostic to the storage entity).

#include "audioscope/dsp/spectrogram.hpp"
#include "audioscope/dsp/fft.hpp"

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <memory>

namespace audioscope {
namespace dsp {

SpectrogramConfig SpectrogramConfig::with_window_size(int size) const {
    SpectrogramConfig c = *this;
    c.window_size = size;
    return c;
}
SpectrogramConfig SpectrogramConfig::with_overlap(int ov) const {
    SpectrogramConfig c = *this;
    c.overlap = ov;
    return c;
}
SpectrogramConfig SpectrogramConfig::with_frequency_range(float min, float max) const {
    SpectrogramConfig c = *this;
    c.min_freq = min;
    c.max_freq = max;
    return c;
}

float SpectrogramData::max_magnitude() const {
    float max = -1e30f;
    for (const auto& row : magnitudes) {
        for (float m : row) if (m > max) max = m;
    }
    return max;
}
float SpectrogramData::min_magnitude() const {
    float min = 1e30f;
    for (const auto& row : magnitudes) {
        for (float m : row) if (m < min) min = m;
    }
    return min;
}

std::vector<std::uint8_t> SpectrogramData::to_image_data(int width, int height) const {
    if (magnitudes.empty() || frequencies.empty()) {
        return std::vector<std::uint8_t>(static_cast<std::size_t>(width) * height * 3, 0);
    }
    const float max_mag = max_magnitude();
    const float min_mag = min_magnitude();
    const float range = std::max(max_mag - min_mag, 0.001f);
    const int time_count = static_cast<int>(time_bins.size());
    const int freq_bins = static_cast<int>(frequencies.size());
    std::vector<std::uint8_t> pixels(static_cast<std::size_t>(width) * height * 3);
    for (int y = 0; y < height; ++y) {
        int freq_idx = static_cast<int>(static_cast<float>(y) / height * freq_bins);
        if (freq_idx > freq_bins - 1) freq_idx = freq_bins - 1;
        for (int x = 0; x < width; ++x) {
            int time_idx = static_cast<int>(static_cast<float>(x) / width * time_count);
            if (time_idx > time_count - 1) time_idx = time_count - 1;
            const float mag = magnitudes[time_idx][freq_idx];
            const int v = static_cast<int>((mag - min_mag) / range * 255.0f);
            const std::uint8_t n = v > 255 ? 255 : (v < 0 ? 0 : static_cast<std::uint8_t>(v));
            auto idx = (static_cast<std::size_t>(y) * width + x) * 3;
            pixels[idx] = n;
            pixels[idx + 1] = static_cast<std::uint8_t>(std::max(n - 50, 0));
            pixels[idx + 2] = static_cast<std::uint8_t>(std::max(n - 100, 0));
        }
    }
    return pixels;
}

class SpectrogramProcessor::Impl {
public:
    FftProcessor fft;
};

SpectrogramProcessor::SpectrogramProcessor() : impl_(std::make_unique<Impl>()) {}
SpectrogramProcessor::~SpectrogramProcessor() = default;

SpectrogramData SpectrogramProcessor::compute(const float* samples, std::size_t count,
                                               float sample_rate, const SpectrogramConfig& config,
                                               std::int64_t start_time_ms) {
    SpectrogramData out;
    out.sample_rate = sample_rate;
    out.window_size = config.window_size;
    out.overlap = config.overlap;
    const int hop = config.window_size - config.overlap;
    if (hop <= 0 || count < static_cast<std::size_t>(config.window_size)) {
        return out;
    }
    std::vector<float> frequencies;
    std::vector<std::vector<float>> magnitudes;
    std::vector<std::int64_t> time_bins;
    std::size_t position = 0;
    while (position + config.window_size <= count) {
        const Spectrum spec = impl_->fft.compute_spectrum(
            samples + position, static_cast<std::size_t>(config.window_size),
            sample_rate, common::WindowType::Hann);
        if (spec.frequencies.empty()) {
            position += hop;
            continue;
        }
        std::size_t f_start = 0;
        for (std::size_t i = 0; i < spec.frequencies.size(); ++i) {
            if (spec.frequencies[i] >= config.min_freq) { f_start = i; break; }
        }
        std::size_t f_end = spec.frequencies.size();
        for (std::size_t i = spec.frequencies.size(); i-- > 0;) {
            if (spec.frequencies[i] <= config.max_freq) { f_end = i + 1; break; }
        }
        std::vector<float> filtered(spec.magnitudes_db.begin() + f_start,
                                    spec.magnitudes_db.begin() + f_end);
        if (frequencies.empty()) {
            frequencies.assign(spec.frequencies.begin() + f_start,
                               spec.frequencies.begin() + f_end);
        }
        const std::int64_t ts = start_time_ms
            + static_cast<std::int64_t>(position) * 1000 / static_cast<std::int64_t>(sample_rate);
        time_bins.push_back(ts);
        magnitudes.push_back(std::move(filtered));
        position += hop;
    }
    out.frequencies = std::move(frequencies);
    out.time_bins = std::move(time_bins);
    out.magnitudes = std::move(magnitudes);
    return out;
}

SpectrogramData SpectrogramProcessor::compute_default(const float* samples, std::size_t count,
                                                       float sample_rate) {
    return compute(samples, count, sample_rate, SpectrogramConfig{});
}

} // namespace dsp
} // namespace audioscope
